third_party/move/move-vm/runtime/src/interpreter.rs

```diff
@@ -1726,83 +1817,60 @@ impl Frame {
-                let args_ty = resolver.get_struct_field_tys(*idx)?;
+                let args_ty = resolver.get_struct(*idx)?;
+                let field_tys = args_ty.fields(None).iter().map(|(_, ty)| ty);
                 let output_ty = resolver.get_struct_ty(*idx);
-                let ability = output_ty.abilities()?;
-
-                // If the struct has a key ability, we expect all of its field to have store ability but not key ability.
-                let field_expected_abilities = if ability.has_key() {
-                    ability
-                        .remove(Ability::Key)
-                        .union(AbilitySet::singleton(Ability::Store))
-                } else {
-                    ability
-                };
-
-                if field_count as usize != args_ty.field_tys.len() {
-                    return Err(
-                        PartialVMError::new(StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR)
-                            .with_message("Args count mismatch".to_string()),
-                    );
-                }
-
-                for (ty, expected_ty) in interpreter
-                    .operand_stack
-                    .popn_tys(field_count)?
-                    .into_iter()
-                    .zip(args_ty.field_tys.iter())
-                {
-                    // Fields ability should be a subset of the struct ability because abilities can be weakened but not the other direction.
-                    // For example, it is ok to have a struct that doesn't have a copy capability where its field is a struct that has copy capability but not vice versa.
-                    ty.paranoid_check_abilities(field_expected_abilities)?;
-                    ty.paranoid_check_eq(expected_ty)?;
-                }
-
-                interpreter.operand_stack.push_ty(output_ty)?;
+                Self::verify_pack(interpreter, field_count, field_tys, output_ty)?;
             },
             Bytecode::PackGeneric(idx) => {
                 let field_count = resolver.field_instantiation_count(*idx);
                 let output_ty = ty_cache.get_struct_type(*idx, resolver, ty_args)?.0.clone();
                 let args_ty = ty_cache.get_struct_fields_types(*idx, resolver, ty_args)?;
-                let ability = output_ty.abilities()?;
-
-                // If the struct has a key ability, we expect all of its field to have store ability but not key ability.
-                let field_expected_abilities = if ability.has_key() {
-                    ability
-                        .remove(Ability::Key)
-                        .union(AbilitySet::singleton(Ability::Store))
-                } else {
-                    ability
-                };
 
                 if field_count as usize != args_ty.len() {
+                    // This is an inconsistency between the cache and the actual
+                    // type declaration. We keep it here for legacy reasons, but
```

Comment:
Do not understand the comment. Why keep for legacy reasons? I would say if it is impossible to hit, we remove it. Otherwise, this is an important invariant to check and we should keep it.

Reply:
I agree with George, still think the comment is not easy to understand (maybe copy-paste your reply above to the comment)

Reply:
Updated the comment